src/hypertable.c

```diff
@@ -848,6 +856,114 @@ hypertable_create_schema(const char *schema_name)
 		);
 }
 
+#define INSERT_BLOCKER_NAME "insert_blocker"
+
+PG_FUNCTION_INFO_V1(hypertable_insert_blocker);
+
+Datum
+hypertable_insert_blocker(PG_FUNCTION_ARGS)
+{
+	TriggerData *trigdata = (TriggerData *) fcinfo->context;
+	const char *relname = get_rel_name(trigdata->tg_relation->rd_id);
+
+	if (!CALLED_AS_TRIGGER(fcinfo))
+		elog(ERROR, "insert_blocker: not called by trigger manager");
+
+	ereport(ERROR,
+			(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+			 errmsg("invalid INSERT on the root table of hypertable \"%s\"", relname),
+			 errdetail("Table \"%s\" is a hypertable and does not support INSERTs on the root table.\n"
+					   "This error typically occurs if the TimescaleDB extension has not been preloaded, or,\n"
+					   "the current session has 'timescaledb.restoring' parameter set to 'ON' for the current session.",
+					   relname)));
+
+	PG_RETURN_NULL();
+}
+
+/*
+ * Get the insert blocker trigger on a table.
+ *
+ * Note that we cannot get the insert trigger by name since internal triggers
+ * are made unique by appending the trigger OID, which we do not
+ * know. Instead, we have to search all triggers.
+ */
+static Oid
+get_insert_blocker_trigger(Oid relid)
+{
+	Relation	tgrel;
+	ScanKeyData skey[1];
+	SysScanDesc tgscan;
+	HeapTuple	tuple;
+	Oid			tgoid = InvalidOid;
+
+	tgrel = heap_open(TriggerRelationId, AccessShareLock);
+
+	ScanKeyInit(&skey[0],
+				Anum_pg_trigger_tgrelid,
+				BTEqualStrategyNumber, F_OIDEQ,
+				ObjectIdGetDatum(relid));
+
+	tgscan = systable_beginscan(tgrel, TriggerRelidNameIndexId, true,
+								NULL, 1, skey);
+
+	while (HeapTupleIsValid(tuple = systable_getnext(tgscan)))
+	{
+		Form_pg_trigger trig = (Form_pg_trigger) GETSTRUCT(tuple);
+
+		if (TRIGGER_TYPE_MATCHES(trig->tgtype, TRIGGER_TYPE_ROW, TRIGGER_TYPE_BEFORE, TRIGGER_TYPE_INSERT) &&
+			strncmp(INSERT_BLOCKER_NAME, NameStr(trig->tgname), strlen(INSERT_BLOCKER_NAME)) == 0 &&
+			trig->tgisinternal)
+		{
+			tgoid = HeapTupleGetOid(tuple);
+			break;
+		}
+	}
+
+	systable_endscan(tgscan);
+	heap_close(tgrel, AccessShareLock);
+
+	return tgoid;
+}
+
+/*
+ * Create an internal trigger that blocks INSERTs on a table.
+ *
+ * The blocking trigger is used to block accidental INSERTs on a hypertable's
+ * root table.
+ */
+static Oid
+add_insert_blocker_trigger(Oid relid)
+{
+	ObjectAddress objaddr;
+	char	   *relname = get_rel_name(relid);
+	Oid			schemaid = get_rel_namespace(relid);
+	char	   *schema = get_namespace_name(schemaid);
+	CreateTrigStmt stmt = {
+		.type = T_CreateTrigStmt,
+		.row = true,
+		.timing = TRIGGER_TYPE_BEFORE,
+		.trigname = INSERT_BLOCKER_NAME,	/* Note, internal triggers get the
+											 * OID appended to the name */
+		.relation = makeRangeVar(schema, relname, -1),
+		.funcname = list_make2(makeString(INTERNAL_SCHEMA_NAME), makeString(INSERT_BLOCKER_NAME)),
```

Comment:
Is this what is forcing you to expose hypertable_insert_blocker at the SQL level?

Reply:
Do we need to create the `stmt` before we check if the trigger already exists?